File: frontend/src/lib/protocol.test.ts
import { describe, expect, it } from 'vitest'

import {
  encodeClientMessage,
  parseServerMessage,
  type ClientMessage,
  type ServerMessage,
} from './protocol'

// ---------- encodeClientMessage round-trip ----------

describe('encodeClientMessage', () => {
  it('encodes operation message', () => {
    const msg: ClientMessage = {
      type: 'operation',
      operation: { kind: 'insert', offset: 3, length: 0, text: 'hi' },
    }
    const encoded = encodeClientMessage(msg)
    const parsed = JSON.parse(encoded) as unknown
    expect((parsed as { type: string }).type).toBe('operation')
    expect((parsed as { operation: { text: string } }).operation.text).toBe('hi')
  })

  it('encodes sync message', () => {
    const encoded = encodeClientMessage({ type: 'sync' })
    const parsed = JSON.parse(encoded) as unknown
    expect((parsed as { type: string }).type).toBe('sync')
  })

  it('encodes presence message', () => {
    const msg: ClientMessage = {
      type: 'presence',
      presence: { anchor: 4, head: 9 },
    }
    const encoded = encodeClientMessage(msg)
    const parsed = JSON.parse(encoded) as { presence: { anchor: number; head: number } }
    expect(parsed.presence.anchor).toBe(4)
    expect(parsed.presence.head).toBe(9)
  })
})

// ---------- parseServerMessage round-trip ----------

describe('parseServerMessage', () => {
  it('parses snapshot message', () => {
    const raw = JSON.stringify({
      type: 'snapshot',
      snapshot: { documentId: 'doc1', version: 2, content: 'hello' },
    })
    const msg = parseServerMessage(raw) as ServerMessage
    expect(msg).not.toBeNull()
    expect(msg.type).toBe('snapshot')
    if (msg.type === 'snapshot') {
      expect(msg.snapshot.content).toBe('hello')
      expect(msg.snapshot.version).toBe(2)
    }
  })

  it('parses presenceSnapshot message', () => {
    const raw = JSON.stringify({
      type: 'presenceSnapshot',
      presences: {
        u1: { userId: 'u1', name: 'Alice', color: '#f00', anchor: 0, head: 1 },
      },
    })
    const msg = parseServerMessage(raw) as ServerMessage
    expect(msg).not.toBeNull()
    expect(msg.type).toBe('presenceSnapshot')
    if (msg.type === 'presenceSnapshot') {
      expect(msg.presences['u1'].name).toBe('Alice')
    }
  })

  it('parses presenceUpdate message', () => {
    const raw = JSON.stringify({
      type: 'presenceUpdate',
      userId: 'u2',
      presence: { userId: 'u2', name: 'Bob', color: '#0f0', anchor: 5, head: 8 },
    })
    const msg = parseServerMessage(raw) as ServerMessage
    expect(msg).not.toBeNull()
    expect(msg.type).toBe('presenceUpdate')
    if (msg.type === 'presenceUpdate') {
      expect(msg.userId).toBe('u2')
      expect(msg.presence.anchor).toBe(5)
    }
  })

  it('parses error message', () => {
    const raw = JSON.stringify({ type: 'error', error: 'something went wrong' })
    const msg = parseServerMessage(raw) as ServerMessage
    expect(msg).not.toBeNull()
    expect(msg.type).toBe('error')
    if (msg.type === 'error') {
      expect(msg.error).toBe('something went wrong')
    }
  })

  it('returns null for unknown type', () => {
    expect(parseServerMessage(JSON.stringify({ type: 'bogus' }))).toBeNull()
  })

  it('returns null for malformed JSON', () => {
    expect(parseServerMessage('not json')).toBeNull()
  })

  it('returns null for missing type field', () => {
    expect(parseServerMessage(JSON.stringify({ data: 'oops' }))).toBeNull()
  })
})

File: backend/internal/realtime/protocol_test.go
package realtime_test

import (
	"encoding/json"
	"testing"
	"time"

	"github.com/boatnoah/notedown/internal/ot"
	"github.com/boatnoah/notedown/internal/realtime"
)

// ---------- UnmarshalClient round-trip tests ----------

func TestUnmarshalClient_Operation(t *testing.T) {
	raw := `{"type":"operation","operation":{"id":"op1","clientId":"c1","kind":"insert","offset":5,"length":0,"text":"hello","timestamp":"0001-01-01T00:00:00Z"}}`
	msg, err := realtime.UnmarshalClient([]byte(raw))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	m, ok := msg.(realtime.OperationMsg)
	if !ok {
		t.Fatalf("got %T, want OperationMsg", msg)
	}
	if m.Operation.Kind != "insert" {
		t.Errorf("kind = %q, want %q", m.Operation.Kind, "insert")
	}
	if m.Operation.Offset != 5 {
		t.Errorf("offset = %d, want 5", m.Operation.Offset)
	}
	if m.Operation.Text != "hello" {
		t.Errorf("text = %q, want %q", m.Operation.Text, "hello")
	}
}

func TestUnmarshalClient_Sync(t *testing.T) {
	msg, err := realtime.UnmarshalClient([]byte(`{"type":"sync"}`))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if _, ok := msg.(realtime.SyncMsg); !ok {
		t.Fatalf("got %T, want SyncMsg", msg)
	}
}

func TestUnmarshalClient_Presence(t *testing.T) {
	raw := `{"type":"presence","presence":{"anchor":3,"head":7}}`
	msg, err := realtime.UnmarshalClient([]byte(raw))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	m, ok := msg.(realtime.PresenceMsg)
	if !ok {
		t.Fatalf("got %T, want PresenceMsg", msg)
	}
	if m.Presence.Anchor != 3 {
		t.Errorf("anchor = %d, want 3", m.Presence.Anchor)
	}
	if m.Presence.Head != 7 {
		t.Errorf("head = %d, want 7", m.Presence.Head)
	}
}

func TestUnmarshalClient_UnknownType(t *testing.T) {
	_, err := realtime.UnmarshalClient([]byte(`{"type":"bogus"}`))
	if err == nil {
		t.Fatal("expected error for unknown type, got nil")
	}
}

// ---------- MarshalServer round-trip tests ----------

func roundTrip(t *testing.T, msg realtime.ServerMsg) map[string]json.RawMessage {
	t.Helper()
	data, err := realtime.MarshalServer(msg)
	if err != nil {
		t.Fatalf("MarshalServer error: %v", err)
	}
	var out map[string]json.RawMessage
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("unmarshal error: %v", err)
	}
	return out
}

func jsonString(t *testing.T, raw json.RawMessage) string {
	t.Helper()
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		t.Fatalf("expected JSON string: %v", err)
	}
	return s
}

func TestMarshalServer_Snapshot(t *testing.T) {
	msg := realtime.SnapshotMsg{Snapshot: ot.Snapshot{
		DocumentID: "doc1",
		Version:    3,
		Content:    "hello world",
	}}
	fields := roundTrip(t, msg)
	if got := jsonString(t, fields["type"]); got != "snapshot" {
		t.Errorf("type = %q, want snapshot", got)
	}

	var snap ot.Snapshot
	if err := json.Unmarshal(fields["snapshot"], &snap); err != nil {
		t.Fatalf("unmarshal snapshot: %v", err)
	}
	if snap.DocumentID != "doc1" || snap.Version != 3 || snap.Content != "hello world" {
		t.Errorf("snapshot fields mismatch: %+v", snap)
	}
}

func TestMarshalServer_PresenceSnapshot(t *testing.T) {
	presences := map[string]realtime.Presence{
		"u1": {UserID: "u1", Name: "Alice", Color: "#ff0000", Anchor: 1, Head: 2, UpdatedAt: time.Time{}},
	}
	fields := roundTrip(t, realtime.PresenceSnapshotMsg{Presences: presences})
	if got := jsonString(t, fields["type"]); got != "presenceSnapshot" {
		t.Errorf("type = %q, want presenceSnapshot", got)
	}
	var out map[string]realtime.Presence
	if err := json.Unmarshal(fields["presences"], &out); err != nil {
		t.Fatalf("unmarshal presences: %v", err)
	}
	if out["u1"].Name != "Alice" {
		t.Errorf("presence name = %q, want Alice", out["u1"].Name)
	}
}

func TestMarshalServer_PresenceUpdate(t *testing.T) {
	msg := realtime.PresenceUpdateMsg{
		UserID:   "u2",
		Presence: realtime.Presence{UserID: "u2", Anchor: 10, Head: 15},
	}
	fields := roundTrip(t, msg)
	if got := jsonString(t, fields["type"]); got != "presenceUpdate" {
		t.Errorf("type = %q, want presenceUpdate", got)
	}
	if got := jsonString(t, fields["userId"]); got != "u2" {
		t.Errorf("userId = %q, want u2", got)
	}
}

func TestMarshalServer_Error(t *testing.T) {
	fields := roundTrip(t, realtime.ErrorMsg{Error: "something went wrong"})
	if got := jsonString(t, fields["type"]); got != "error" {
		t.Errorf("type = %q, want error", got)
	}
	if got := jsonString(t, fields["error"]); got != "something went wrong" {
		t.Errorf("error = %q, want %q", got, "something went wrong")
	}
}
